dbms/src/TiDB/Schema/SchemaBuilder.cpp

```diff
@@ -410,70 +460,70 @@ void SchemaBuilder<Getter, NameMapper>::applyPartitionDiff(
     const TableInfoPtr & table_info,
     const ManageableStoragePtr & storage)
 {
-    const auto & orig_table_info = storage->getTableInfo();
-    if (!orig_table_info.isLogicalPartitionTable())
+    const auto & local_table_info = storage->getTableInfo();
+    // ALTER TABLE t PARTITION BY ... may turn a non-partition table into partition table
+    // with some partition ids in `partition.adding_definitions`/`partition.definitions`
+    // and `partition.dropping_definitions`. We need to create those partitions.
+    if (!local_table_info.isLogicalPartitionTable())
     {
-        LOG_ERROR(
+        LOG_INFO(
             log,
-            "old table in TiFlash not partition table {} with database_id={}, table_id={}",
-            name_mapper.debugCanonicalName(*db_info, orig_table_info),
+            "Altering non-partition table to be a partition table {} with database_id={}, table_id={}",
+            name_mapper.debugCanonicalName(*db_info, local_table_info),
             db_info->id,
-            orig_table_info.id);
-        return;
+            local_table_info.id);
     }
 
-    const auto & orig_defs = orig_table_info.partition.definitions;
+    const auto & local_defs = local_table_info.partition.definitions;
     const auto & new_defs = table_info->partition.definitions;
 
-    std::unordered_set<TableID> orig_part_id_set, new_part_id_set;
-    std::vector<String> orig_part_ids, new_part_ids;
-    std::for_each(orig_defs.begin(), orig_defs.end(), [&orig_part_id_set, &orig_part_ids](const auto & def) {
-        orig_part_id_set.emplace(def.id);
-        orig_part_ids.emplace_back(std::to_string(def.id));
+    std::unordered_set<TableID> local_part_id_set, new_part_id_set;
+    std::for_each(local_defs.begin(), local_defs.end(), [&local_part_id_set](const auto & def) {
+        local_part_id_set.emplace(def.id);
     });
-    std::for_each(new_defs.begin(), new_defs.end(), [&new_part_id_set, &new_part_ids](const auto & def) {
+    std::for_each(new_defs.begin(), new_defs.end(), [&new_part_id_set](const auto & def) {
         new_part_id_set.emplace(def.id);
-        new_part_ids.emplace_back(std::to_string(def.id));
     });
 
-    auto orig_part_ids_str = boost::algorithm::join(orig_part_ids, ", ");
-    auto new_part_ids_str = boost::algorithm::join(new_part_ids, ", ");
-
     LOG_INFO(
         log,
         "Applying partition changes {} with database_id={}, table_id={}, old: {}, new: {}",
         name_mapper.debugCanonicalName(*db_info, *table_info),
         db_info->id,
         table_info->id,
-        orig_part_ids_str,
-        new_part_ids_str);
+        local_part_id_set,
+        new_part_id_set);
```

Comment:
Can we use `std::unordered_set` in `fmt::format` directly?

Reply:
Yes, fmtlib can handle it after includeing the `<fmt/ranges>` (included in the logger_useful.h)